query-translator/tests/translate.rs

```diff
@@ -550,4 +550,125 @@ fn test_complex_nested_or_join_type_projection() {
+
+    // Verify that we accept inline constants.
+    let query = r#"[:find ?x ?y :where [(ground $ [1 "yyy"]) [?x ?y]]]"#;
+    let SQLQuery { sql, args } = translate(&schema, query);
+    assert_eq!(sql, "SELECT DISTINCT `c00`.`?x` AS `?x`, `c00`.`?y` AS `?y` FROM \
+                         (SELECT 0 AS `?x`, 0 AS `?y` WHERE 0 UNION ALL VALUES (1, $v0)) AS `c00`");
+    assert_eq!(args, vec![make_arg("$v0", "yyy")]);
+
+    // Verify that we accept bound input constants.
+    let query = r#"[:find [?x ?y] :in ?u ?v :where [(ground $ [?u ?v]) [?x ?y]]]"#;
+    let inputs = QueryInputs::with_value_sequence(vec![(Variable::from_valid_name("?u"), TypedValue::Long(2)),
+                                                       (Variable::from_valid_name("?v"), TypedValue::String(Rc::new("aaa".into()))),]);
+    let SQLQuery { sql, args } = translate_with_inputs(&schema, query, inputs);
+    // TODO: treat 2 as an input variable that could be bound late, rather than eagerly binding it.
+    assert_eq!(sql, "SELECT `c00`.`?x` AS `?x`, `c00`.`?y` AS `?y` FROM \
+                         (SELECT 0 AS `?x`, 0 AS `?y` WHERE 0 UNION ALL VALUES (2, $v0)) AS `c00` LIMIT 1");
+    assert_eq!(args, vec![make_arg("$v0", "aaa"),]);
+}
+
+#[test]
+fn test_ground_coll() {
+    let schema = prepopulated_schema();
+
+    // Verify that we accept inline constants.
+    let query = r#"[:find ?x :where [(ground $ ["xxx" "yyy"]) [?x ...]]]"#;
+    let SQLQuery { sql, args } = translate(&schema, query);
+    assert_eq!(sql, "SELECT DISTINCT `c00`.`?x` AS `?x` FROM \
+                         (SELECT 0 AS `?x` WHERE 0 UNION ALL VALUES ($v0), ($v1)) AS `c00`");
+    assert_eq!(args, vec![make_arg("$v0", "xxx"),
+                          make_arg("$v1", "yyy")]);
+
+    // Verify that we accept bound input constants.
+    let query = r#"[:find ?x :in ?u ?v :where [(ground $ [?u ?v]) [?x ...]]]"#;
+    let inputs = QueryInputs::with_value_sequence(vec![(Variable::from_valid_name("?u"), TypedValue::Long(2)),
+                                                       (Variable::from_valid_name("?v"), TypedValue::Long(3)),]);
+    let SQLQuery { sql, args } = translate_with_inputs(&schema, query, inputs);
+    // TODO: treat 2 and 3 as input variables that could be bound late, rather than eagerly binding.
+    assert_eq!(sql, "SELECT DISTINCT `c00`.`?x` AS `?x` FROM \
+                         (SELECT 0 AS `?x` WHERE 0 UNION ALL VALUES (2), (3)) AS `c00`");
+    assert_eq!(args, vec![]);
+}
+
+#[test]
+fn test_ground_rel() {
+    let schema = prepopulated_schema();
+
+    // Verify that we accept inline constants.
+    let query = r#"[:find ?x ?y :where [(ground $ [[1 "xxx"] [2 "yyy"]]) [[?x ?y]]]]"#;
+    let SQLQuery { sql, args } = translate(&schema, query);
+    assert_eq!(sql, "SELECT DISTINCT `c00`.`?x` AS `?x`, `c00`.`?y` AS `?y` FROM \
+                         (SELECT 0 AS `?x`, 0 AS `?y` WHERE 0 UNION ALL VALUES (1, $v0), (2, $v1)) AS `c00`");
```

Comment:
Ooh, tricky. What is the type of `1`? Do we expect the projector to give you `Ref` or `Long`?